Add unit tests and code coverage

rwalton-arm opened the pull request:
This PR makes the following changes:

Adds unit tests using GoogleTest. GoogleTest will be fetched automatically by CMake at 
configuration-time.
To stop CMake from trying to fetch GoogleTest pass the `-DBUILD_TESTING=OFF`
option to CMake when configuring the build system.

Adds an ENABLE_COVERAGE cmake option, when this is ON then the coverage flags
are added to the compiler and linker options. A job is added to the github CI workflow 
to run the unit tests and upload results to codecov.

Reworks the cmake targets to remove the option for disabling stdio. 
Instead we now have a "client_userio" target, so consumers can just link to the target
they need depending on their IO setup. This is easier for users than trying to forward 
options to greentea-client at configure time to disable stdio, which may not even be 
feasible if the user isn't building greentea-client from source.

rwalton-arm committed:
cmake: Use `option` for user settable configuration

Previously we exposed an interface of cache variables allowing the user
to select different build configurations or build example applications.
However, it is more semantically correct to use `option` variables for
these configuration settings. Cache variables are usually understood to
be internal implementation details that can't be safely changed without
the risk of side effects. `option` variables are designed to be
user-facing, are limited to BOOL values, and are better supported in
ccmake and cmake-gui.

rwalton-arm committed:
cmake: Enable code coverage

Add an ENABLE_COVERAGE option to enable code coverage.

rwalton-arm committed:
cmake: Require CMake 3.14 or later

CMake 3.14 introduced FetchContent_MakeAvailable, which is a useful
helper for making external content available in the main build.

rwalton-arm committed:
cmake: Enable unit testing with GoogleTest

Add a CMakeLists.txt for enabling unit testing using GoogleTest. The
CMakeLists.txt contains an executable `greentea-tests` target depending
on greentea-client and GoogleTest. Unit tests for greentea-client's
KiVi protocol send and receive functions will be added to
test_kv_protocol.cpp, which is a source file added to the executable
target.

GoogleTest will be fetched automatically by CMake at configuration-time.
To stop CMake from trying to fetch GoogleTest pass the `-DBUILD_TESTING=OFF`
option to CMake when configuring the build system.

rwalton-arm committed:
Add fake console IO fixture for unit testing

Add a "fake console" implementation of greentea-client's IO for testing.
The fake implementation models console stdout and stdin as static string
buffers. An interface is provided to set stdin and retrieve stdout from
the fake console.

rwalton-arm committed:
Add unit tests for greentea_send_kv

Add unit tests for greentea_send_kv using the Console fixture.

rwalton-arm committed:
Add tests for greentea_test_env API

Add unit tests for greentea-client's test flow control and reporting
functions.

rwalton-arm committed:
ci: Upload coverage to codecov

Add a workflow step to upload code coverage results to codecov.

rwalton-arm committed:
cmake: Use different targets to select IO types

Remove the option to select IO type and use another CMake target
instead. This allows more flexibility for users and means they can just
link to the target they need to build instead of trying to forward
option flags to greentea-client, which may not always be feasible if not
building the client from source.

rwalton-arm commented on CMakeLists.txt:
```diff
@@ -11,28 +11,36 @@ project(greentea
+)
+target_include_directories(client
+    PUBLIC
+        "$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>"
```
This duplication of include directories and source files isn't ideal. When I tried to inherit the "client_userio" target it caused errors in the CI (but not locally on my macbook using clang or gcc10 or 11) due to missing definitions for the greentea_putc and greentea_getc IO functions. I suspect it was due to link order which somehow the mac versions of GCC and clang were masking.

If someone can suggest a better solution for this, which doesn't cause the link errors, it would be useful.

LDong-Arm replied:
How did you inherit it? Linking one to the other?

LDong-Arm replied:
Yea I can reproduce it on Ubuntu...

LDong-Arm replied:
I guess you did something like
```cmake
add_library(client
    source/greentea_test_io.c
)

target_link_libraries(client
    PUBLIC
        client_userio
)
```
?

What happens is, `libclient.a` will only contain greentea_test_io without any contents of `libclient_userio.a`. Instead, when you build an example, CMake will link both both static libraries to the executable, probably something like `-lclient -lclient_userio`.

Maybe the clang linker doesn't care about the order, or simply fuses all the symbols to the executable and only cleans up unused ones at the very end. This is also our experience with armclang. GCC seems to only link what's needed at each step I guess (from my experience with GCC_ARM but might apply to host GCC too).

The linker first links `libclient.a` and decides there's nothing useful, because main() doesn't call any of the the IO function. Then when it proceeds to link `libclient_userio.a`, it can't find the IO functions.

LDong-Arm replied:
In terms of API offering, linking `client` to `client_userio` makes sense. But in terms code dependency, it doesn't. The IO functions in the former are actually called by the latter, so it's a dependency reversal in a sense.

Maybe we could explore a way to fuse `libclient_userio.a` into `libclient.a`, so `libclient.a` will become a self-contained, shippable library with the whole greentea-client instead of just the IO part.

rwalton-arm replied:
Yeah, weirdly with GCC 10 and GCC 11 on my macbook it also linked without issue. It's only when I ran it in the CI, then tried it locally on ArchLinux where the link order became a problem. I'll try and find a way to reverse the dependencies as you suggested.

rwalton-arm replied:
I couldn't really think of a nicer way of doing this without resorting to some INTERFACE library hacking or using variables. I'm fine with leaving it how it is I think. It's a small amount of duplication and in some ways it's probably easier to grok for someone new to CMake.

If there's an actual issue with this then I can spend a bit more time on it. I think I've probably used up enough time on this now though!

LDong-Arm replied:
I'm also fine with it. Another idea is to make the IO functions weak symbols, but I roughtly remember we would avoid it in the future, possibly to avoid uncertainty in what gets linked.

Patater replied:
I don't think weak symbols will help, given they get hard resolved before going into a `.a` file. As I understand it, you can't override a symbol (even if you put "weak" in the C file) in a `.a` file with a stronger symbol from another `.a` file: because after the `.a` is made, all symbols are "strong".

Patater commented on tests/doubles/fake_console_io.cpp:
```diff
@@ -0,0 +1,64 @@
+int greentea_getc()
+{
+    int out = EOF;
+    if(!_stdin.empty())
```
Do we not put a space after `if`? Is astyle not catching this?

rwalton-arm replied:
I'll take a look. I also thought it was slightly odd that astyle didn't flag anything. The astyle check is a copy-paste from the mbed-os travis job, I'm assuming it worked there.

rwalton-arm replied:
I've fixed the astyle check (it's now failing as expected). Now to fix the formatting!

Patater commented on tests/unit/CMakeLists.txt:
```diff
@@ -13,7 +13,8 @@ FetchContent_Declare(
 add_executable(greentea-tests test_kv_protocol.cpp)
-target_link_libraries(greentea-tests PRIVATE greentea::client gtest_main)
+target_compile_features(greentea-tests PUBLIC cxx_std_14)
+target_link_libraries(greentea-tests PUBLIC greentea::client_userio fake-console-io gtest_main)
```
Was `greentea-tests` the target that failed linking in CI? Did you try in CI with listing `fake-console-io` first here?

rwalton-arm replied:
No, it was the example apps that failed linking. I'll spend another hour trying to figure it out, it probably was something silly like that.

Patater commented on tests/unit/test_kv_protocol.cpp:
```diff
@@ -96,3 +96,57 @@ TEST_F(KiViProtocolTest, SendStringValueAndPassFailCount)
+    std::string console = fake_console.get_stdout();
+    ASSERT_NE(console.find(GREENTEA_TEST_ENV_END), std::string::npos);
+    ASSERT_NE(console.find(GREENTEA_TEST_ENV_SUCCESS), std::string::npos);
+    ASSERT_NE(console.find(GREENTEA_TEST_ENV_EXIT), std::string::npos);
```
For a lot of console output, doesn't order matter?

For example, shouldn't we fail the test if EXIT is sent before SUCCESS?

rwalton-arm replied:
Good point, will add assertions to ensure the order is correct.

rwalton-arm committed:
ci: Fix astyle check

The astyle check wasn't working because github actions only check out
with depth=1 in a detached HEAD state. Causing the HEAD..FETCH_HEAD diff
to not return any changed files.

Now we just run astyle on all relevant files in the source tree, rather
than trying to check only the files changed in the PR.

rwalton-arm committed:
astyle: Fix formatting in greentea_test_io.c

LDong-Arm commented on tests/unit/CMakeLists.txt:
```diff
@@ -0,0 +1,17 @@
+FetchContent_Declare(
+  googletest
+  GIT_REPOSITORY https://github.com/google/googletest.git
+  GIT_TAG        master
```
Is it possible to use a release?

rwalton-arm replied:
googletest [recommend using the latest master](https://github.com/google/googletest#live-at-head), they don't really make releases anymore. They're following the "Abseil Live at HEAD" philosophy, which means they make certain "promises" about API compatibility and not breaking downstream code. All the older tagged release versions of googletest also produce some CMake configuration warnings:

```
    CMake Deprecation Warning at __build/_deps/googletest-src/CMakeLists.txt:4 (cmake_minimum_required):
      Compatibility with CMake < 2.8.12 will be removed from a future version of
      CMake.

      Update the VERSION argument <min> value or use a ...<max> suffix to tell
      CMake that the project does not need compatibility with older versions.

    CMake Deprecation Warning at __build/_deps/googletest-src/googlemock/CMakeLists.txt:45 (cmake_minimum_required):
      Compatibility with CMake < 2.8.12 will be removed from a future version of
      CMake.

      Update the VERSION argument <min> value or use a ...<max> suffix to tell
      CMake that the project does not need compatibility with older versions.

    CMake Deprecation Warning at __build/_deps/googletest-src/googletest/CMakeLists.txt:56 (cmake_minimum_required):
      Compatibility with CMake < 2.8.12 will be removed from a future version of
      CMake.

      Update the VERSION argument <min> value or use a ...<max> suffix to tell
      CMake that the project does not need compatibility with older versions.
```

LDong-Arm replied:
Cool, thanks

LDong-Arm commented on tests/unit/CMakeLists.txt:
```diff
@@ -15,3 +15,18 @@ FetchContent_MakeAvailable(googletest)
+
+# Coverage
+
+option(ENABLE_COVERAGE "Enable code coverage" OFF)
```
Why is it off by default?

Patater replied:
It requires lowering the optimization level and causes extra .gcno type files to get generated when run.

LDong-Arm replied:
The unit tests are only built and run on a PC and serves to test functionality instead of performance, so it might be less of a problem?

rwalton-arm replied:
I thought users would prefer to opt in to creating the coverage files. We can change it to ON by default if we think that's preferable?

LDong-Arm replied:
In terms of use cases, if unit testing alone is more common than coverage, then I'm personally fine with keeping it OFF.

LDong-Arm reviewed (commented):
@rwalton-arm Thanks, looks really good to me
